### ISSUE TITLE
pkg.refresh_db not working 

### ISSUE THREAD

**Issue by t-bhard** [saltstack/salt#37669]
### Description of Issue/Question
For some reason pkg.refresh_db fails on my minion and I don't know why this is happening. My setup is extremely straight forward. The weirdest thing too was that it worked fine and when I came over the weekend it just stopped working and i don't know what I did to break it. 

### Setup
All right so let's start with the error it throws when i run pkg.refresh_db:
```
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Including configuration from '/etc/salt/master.d/master.conf'
[DEBUG   ] Reading configuration from /etc/salt/master.d/master.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: ECS-141abdb2.ecs.ads.autodesk.com
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Including configuration from '/etc/salt/master.d/master.conf'
[DEBUG   ] Reading configuration from /etc/salt/master.d/master.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: ECS-141abdb2.ecs.ads.autodesk.com
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'ECS-141abdb2.ecs.ads.autodesk.com_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/master/master_event_pub.ipc
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] Reading minion list from /var/cache/salt/master/jobs/37/32a84ef669f42a36e76c07f738738c/.minions.p
[DEBUG   ] get_iter_returns for jid 20161114145741459046 sent to set(['dss']) will timeout at 14:57:46.468061
[DEBUG   ] Checking whether jid 20161114145741459046 is still running
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'ECS-141abdb2.ecs.ads.autodesk.com_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Passing on saltutil error. This may be an error in saltclient. 'retcode'
[DEBUG   ] Checking whether jid 20161114145741459046 is still running
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'ECS-141abdb2.ecs.ads.autodesk.com_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Passing on saltutil error. This may be an error in saltclient. 'retcode'
[DEBUG   ] jid 20161114145741459046 return from dss
[DEBUG   ] LazyLoaded nested.output
dss:
The minion function caused an exception: Traceback (most recent call last):
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 1332, in _thread_return
    return_data = executor.execute()
  File "C:\salt\bin\lib\site-packages\salt\executors\direct_call.py", line 28, in execute
    return self.func(*self.args, **self.kwargs)
  File "C:\salt\bin\lib\site-packages\salt\modules\win_pkg.py", line 376, in refresh_db
    genrepo(saltenv=saltenv)
  File "C:\salt\bin\lib\site-packages\salt\modules\win_pkg.py", line 431, in genrepo
    for version, repodata in six.iteritems(versions):
  File "C:\salt\bin\lib\site-packages\salt\ext\six.py", line 583, in iteritems
    return iter(d.iteritems(**kw))
AttributeError: 'str' object has no attribute 'iteritems'
[DEBUG   ] jid 20161114145741459046 found all minions set(['dss'])
```

This is how my master config file (etc/salt/minion) looks like:

```

state_top: Win10.sls

 The location of the top file
 file_roots:
 base:
    - /srv/salt/

 winrepo_dir: /srv/salt/win/repo-ng

fileserver_backend:
  - git
  - roots

winrepo_dir_ng: '/srv/salt/win/repo-ng'
winrepo_provider: gitpython

```

My repo did have custom definitions for programs which I thought may have caused the issue. I backed up the folders and completely cleared my repo and repo-ng folders. 

I then ran both:
```
sudo salt-run winrepo.update_git_repos
sudo salt-run winrepo.genrepo

```
The process was successful but running pkg.refresh_db failed nonetheless. I tried uninstalling cygwin from my minion (which was what I was installing during all this) and it still does not work. 

### Versions Report
```
Salt Version:
           Salt: 2016.3.4

Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.2.2
       dateutil: 1.4.1
          gitdb: 0.6.4
      gitpython: 2.0.8
          ioflo: Not Installed
         Jinja2: 2.2.1
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.20.2
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.6.6 (r266:84292, Aug  9 2016, 06:11:56)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.3.1
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 3.2.5

System Versions:
         dist: redhat 6.6 Santiago
        machine: x86_64
        release: 2.6.32-504.12.2.el6.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 6.6 Santiago
```

**Comment by dahrens** [comment]
I did some investigations on this issue and posted an answer on [stackoverflow](http://stackoverflow.com/questions/40592290/saltclient-error/40598367) - if I am right with my assumptions this issue is related to win_pkg which iterates over all sls files, expecting versions to be a dictionary, but does not prove it, before handing this over to six - six dies than, as it can not call `iteritems()` on the string.

I guess this error should be catched in the module.

**Comment by Ch3LL** [comment]
Thanks @dahrens for the great investigation work!

@t-bhard any way you can look through your files and and check through versions as stated on that stack overflow issue? Thanks

**Comment by damon-atkins** [comment]
Try 2016.11 pre-release   it now reports back issues with the files, including structural issues.
When I did this, I did not target old releases as its a behaviour change.

**Comment by t-bhard** [comment]
@dahrens Your assumption seems to be on the nose.
 What I did was, like suggested to me by you on stack, was run salt locally (did require installing MS Visual C++ 2008 Redistributable -x64).

`salt-call pkg.refresh_db -l debug`

Once it completed it gave me the same error as on the master but with a tad more information. Looks like in my case it was with the cygwin's init.sls file located under 

`C:\salt\var\cache\salt\minion\files\base\win`

I ended wanting to start from scratch do i removed everything there. Ran refresh_db locally again and it was a success. I then tried running it from the master and received the exception that had something to do with IOloops (disappeared after a minute) and after that the following error

` Exception: DummyFuture does not support blocking for results`

To get rid of this i headed to Services in Windows and simply stopped the 'salt-minion' service and started it again. 

After that worked like a charm!

Thanks @dahrens  @Ch3LL  @damon-atkins
